src/native/eventpipe/ds-eventpipe-protocol.c

```diff
@@ -336,6 +370,62 @@ ds_eventpipe_collect_tracing2_command_payload_free (EventPipeCollectTracing2Comm
 	ep_rt_object_free (payload);
 }
 
+/*
+* EventPipeCollectTracing3CommandPayload
+*/
+
+static
+uint8_t *
+eventpipe_collect_tracing3_command_try_parse_payload (
+	uint8_t *buffer,
+	uint16_t buffer_len)
+{
+	EP_ASSERT (buffer != NULL);
+
+	uint8_t * buffer_cursor = buffer;
+	uint32_t buffer_cursor_len = buffer_len;
+
+	EventPipeCollectTracing3CommandPayload *instance = ds_eventpipe_collect_tracing3_command_payload_alloc ();
+	ep_raise_error_if_nok (instance != NULL);
+
+	instance->incoming_buffer = buffer;
+
+	if (!eventpipe_collect_tracing_command_try_parse_circular_buffer_size (&buffer_cursor, &buffer_cursor_len, &instance->circular_buffer_size_in_mb ) ||
+		!eventpipe_collect_tracing_command_try_parse_serialization_format (&buffer_cursor, &buffer_cursor_len, &instance->serialization_format) ||
+		!eventpipe_collect_tracing_command_try_parse_rundown_requested (&buffer_cursor, &buffer_cursor_len, &instance->rundown_requested) ||
+		!eventpipe_collect_tracing_command_try_parse_stacks_requested (&buffer_cursor, &buffer_cursor_len, &instance->stacks_requested) ||
+		!eventpipe_collect_tracing_command_try_parse_config (&buffer_cursor, &buffer_cursor_len, &instance->provider_configs))
+		ep_raise_error ();
+
+ep_on_exit:
+	return (uint8_t *)instance;
+
+ep_on_error:
+	ds_eventpipe_collect_tracing3_command_payload_free (instance);
+	instance = NULL;
+	ep_exit_error_handler ();
+}
+
+EventPipeCollectTracing3CommandPayload *
+ds_eventpipe_collect_tracing3_command_payload_alloc (void)
+{
+	return ep_rt_object_alloc (EventPipeCollectTracing3CommandPayload);
+}
+
+void
+ds_eventpipe_collect_tracing3_command_payload_free (EventPipeCollectTracing3CommandPayload *payload)
+{
+	ep_return_void_if_nok (payload != NULL);
```

Comment:
I believe ds_eventpipe_collect_tracing_command_payload_free, ds_eventpipe_collect_tracing2_command_payload_free and ds_eventpipe_collect_tracing3_command_payload_free are identical and should share the same underlying implementation. If we make collect3 into something different, then we could at least make this free implementation shared by ds_eventpipe_collect_tracing_command_payload_free, ds_eventpipe_collect_tracing2_command_payload_free.